### PR TITLE
GSdx ogl: SSO Workaround for AMD buggy drivers

### LINKED ISSUE #1343: [AMD GPU] Serious graphics problems all games opengl-hw recent builds
There's a bug that was only recent in newer builds. i Used to have blending accuracy on high and everything worked fine but now sometime since build pcsx2-v1.5.0-dev-511-gf1b3d7d. i cant play any games really unless i put it on ultra and that makes it painfully slow So i know for sure there is a error there because it used to work just fine until recently. I have a 8350fx and a 7970 xfx card all drivers up to date as well. ill post images as well of what all games do now on the newer builds unless i go ultra.
blending unit error
ultra

### PR DESCRIPTION
**Pull request by lightningterror**
Added Greg's SSO Workaround for AMD buggy drivers with some additions. Fixes #1343 
All 2017 drivers are now blacklisted.
The BSOD/crash issue is still there so don't set Blending Accuracy to None! Check #1552

### CODE REVIEW

**Comment by mirh** [comment]
A better naming for what you are doing is "expand blacklist of AMD driver". 
As I was kind of saying in #1552 though.. What's the point in separate fglrx and legacy_fglrx ?

**Comment by gregory38** [comment]
The initial plan was to have a half broken legacy driver and a good driver. I didn't expect that supposed driver will be a completely broken driver.... Seriously it would be faster and honest for them to say that they don't support openGL

**Comment by mirh** [comment]
Oh, ok, so there's no difference at least atm. 

You could always put 16.2 and 16.3 to fglrx_buggy, which in turns only just disable "none" accuracy though.

**Comment by gregory38** [comment]
```
which in turns only just disable "none" accuracy though.
```
I don't remember we have some code that force BA for anyone.

**Comment by lightningterror** [comment]
Even with that there are still issues , when using OpenGL the emulator crashes a lot.

**Comment by mirh** [comment]
*stares at coders*

**Comment by lightningterror** [comment]
On GT4 I seem to get them at certain parts depending on the maps.

File: plugins/GSdx/GLLoader.cpp [hunk 1]
```diff
@@ -335,12 +335,12 @@ namespace GLLoader {
 		fprintf(stdout, "OpenGL information. GPU: %s. Vendor: %s. Driver: %s\n", glGetString(GL_RENDERER), vendor, &s[v]);
 
 		// Name changed but driver is still bad!
-		if (strstr(vendor, "ATI") || strstr(vendor, "Advanced Micro Devices"))
+		if (strstr(vendor, "ATI") || strstr(vendor, "Advanced Micro Devices") || strstr(vendor, "ATI Technologies Inc."))
```
Comment (FlatOutPS2) [inline plugins/GSdx/GLLoader.cpp]: Very minor note: It would be nicer if `"ATI"` and `"ATI Technologies Inc."` were next to each other. :P

Comment (lightningterror) [inline plugins/GSdx/GLLoader.cpp]: Done however I won't push it yet , I want to make some other changes.

I noticed the gs version is 1.1.0 shown in the emulator , however checking the dll file version and product version it's 1.0.1.9  
Should this be updated ?

**Comment by lightningterror** [comment]
@FlatOutPS2 Done.
@mirh Here is the error log.
[errorlog.txt](https://github.com/PCSX2/pcsx2/files/845788/errorlog.txt)

**Comment by mirh** [comment]
Tbh, windgb report of the memory dump would have been way more helpful. 
And I just tried very briefly GT4 (slideshows are hard to bear) and it seemed fine

**Comment by lightningterror** [comment]
You mean this? 
`Unhandled` exception at 0x3000348D in pcsx2.exe: 0xC0000005: Access violation reading location 0x0034D203

If there is a handler for this exception, the program may be safely continued..

**Comment by mirh** [comment]
I mean, attach windbg to pcsx2, then when it crashes it *should* prompt you.

**Comment by lightningterror** [comment]
With or without Heap Information ? With heap it's about 1gb which will take a while to upload.

**Comment by gregory38** [comment]
You're kidding me. I want touch AMD/win stuff even      with a 10 foot pole.

**Comment by lightningterror** [comment]
Greg right now xD
http://i.imgur.com/ExBLAI3.jpg

Anyway the PR is ready for merge.

**Comment by mirh** [comment]
> Greg right now xD

You miss the Windows logo. 

> Anyway the PR is ready for merge.

*had suggested improvements*

**Comment by gregory38** [comment]
Yes Seriously ! Where is the TROLL and the click bait.

**Comment by gregory38** [comment]
Is it fine for everybody ?

**Comment by mirh** [comment]
In the end I guess those that will keep <17.2 drivers just for this will be on the same order of magnitude of those knowing and using `override_GL_ARB_separate_shader_objects` for testing regardless.. So, whatever.

**Comment by FlatOutPS2** [comment]
We should do something about the messages printed for OpenGL. Every time you switch renderer(especially annoying when switching between OGL SW and HW with the F9 key), it prints more than 15 lines of OpenGL information in the log on Windows. Whereas on Linux, the driver information is only printed once when first booting PCSX2.

**Comment by MrCK1** [comment]
Maybe don't print GL extensions unless they're detected as unsupported?

**Comment by gregory38** [comment]
What are the 15 lines printed?

**Comment by FlatOutPS2** [comment]
> What are the 15 lines printed?

All the extensions and other normal driver information are printed every time, and then there's this:

```
The OpenGL hardware renderer is slow on AMD GPUs due to an inefficient driver. Check out the links below for further information.
https://community.amd.com/message/2756964
https://community.amd.com/thread/205702
Note: Due to an AMD OpenGL driver issue, setting Blending Unit Accuracy to "None" can cause an application or system crash.
Keep Blending Unit Accuracy set to at least the default "Basic" level.
AMD has a fix for the issue that will be released in the coming months. The issue does not affect AMD GPUs on legacy drivers.
```

**Comment by mirh** [comment]
I'd say that *even now*, chances are most people still wouldn't notice it. 
Also, tbh.. Is that really all of that problem if you aren't a *compulsive tester*?

**Comment by FlatOutPS2** [comment]
> Also, tbh.. Is that really all of a problem if you aren't a compulsive tester?

Yes, because it's so much information it can actually hide important messages printed in the log.

**Comment by lightningterror** [comment]
I would also recommend shortening the message to something like this.
```
The OpenGL hardware render is slow on AMD GPUs due to an inefficient driver. Check out the link for further information.
https://github.com/PCSX2/pcsx2/issues/1552
```

And also editing the first post of the git issue with all the needed information required.

**Comment by mirh** [comment]
> Yes, because it's so much information it can actually hide important messages printed in the log.

Ok, then it's only a problem for testers 😛 

> And also editing the first post of the git issue with all the needed information required.

A [wiki](https://github.com/PCSX2/pcsx2/wiki) page would probably fit the purpose better.

**Comment by lightningterror** [comment]
> A wiki page would probably fit the purpose better.

That's even better.

**Comment by FlatOutPS2** [comment]
>Ok, then it's only a problem for testers 😛

It could also make regular users miss some useful log information too. In any case it working like the Linux version would be ideal.

> Check out the link for further information.

I'd leave the `below` in there.

**Comment by lightningterror** [comment]
Okay I created the wiki page. Feel free to edit the info since I know there is a lot of stuff I missed probably.

Anyway moving back on topic how does this sound?
```
Detected inefficient AMD OpenGL driver. Check the link below for further information.
https://github.com/PCSX2/pcsx2/wiki/OpenGL-and-AMD-GPUs---All-you-need-to-know
```

Or the old one
```
The OpenGL hardware render is slow on AMD GPUs due to an inefficient driver. Check out the link for further information.
https://github.com/PCSX2/pcsx2/wiki/OpenGL-and-AMD-GPUs---All-you-need-to-know
```

**Comment by FlatOutPS2** [comment]
The old one is better. The new one is very ambiguous.

**Comment by mirh** [comment]
For the moment (EVERY amd driver sucks), I'd keep the later. 
When they'll finally have a working one.. "detected bad version" will be better. 

Also, change inefficient to buggy.

**Comment by gregory38** [comment]
A working driver might never happen (quite likely)

**Comment by FlatOutPS2** [comment]
> A working driver might never happen (quite likely)

It will happen, but likely not the driver made by AMD.

Besides, once they fix the current BSOD issue and the recent regression, and GSdx gets multi-threading it should already be much better.

**Comment by gregory38** [comment]
Not sure what you mean by "not by AMD". The opensource driver is fine. It will even get MT support once I got back from my vacation to finish some patches.

**Comment by mirh** [comment]
If you are so fervent on mesa, the HD 6310 on my laptop is always ready to be ssh'd. 😄

**Comment by gregory38** [comment]
@mirh I can run `nouveau` on my Nvidia GPU. I alternate between the 2 drivers when I reboot but I barely reboot my computer those days.

So are we fine with this PR ?

**Comment by mirh** [comment]
I was actually more referring to your goodwillingness to work on mesa, rather than pcsx2 specifically 
nvc0 is almost too boring to develop now, while r600 has still a long road ahead :p

Anyway, yes fine for me. 
Pretty nice you don't have to make a PR to edit the wiki entry.

**Comment by gregory38** [comment]
I think r600 is fine with PCSX2. My goal is to have a good experience with PCSX2 on Mesa not to provide a nice driver for everyone ;)

**Comment by gregory38** [comment]
@FlatOutPS2 is it ok for you ?

File: plugins/GSdx/GLLoader.cpp [hunk 2]
```diff
@@ -437,12 +437,8 @@ namespace GLLoader {
 		status &= status_and_override(found_GL_ARB_get_texture_sub_image, "GL_ARB_get_texture_sub_image");
 
 		if (fglrx_buggy_driver) {
-			fprintf(stderr, "The OpenGL hardware renderer is slow on AMD GPUs due to an inefficient driver. Check out the links below for further information.\n"
-					"https://community.amd.com/message/2756964\n"
-					"https://community.amd.com/thread/205702\n"
-					"Note: Due to an AMD OpenGL driver issue, setting Blending Unit Accuracy to \"None\" can cause an application or system crash.\n"
-					"Keep Blending Unit Accuracy set to at least the default \"Basic\" level.\n"
-					"AMD has a fix for the issue that will be released in the coming months. The issue does not affect AMD GPUs on legacy drivers.\n");
+			fprintf(stderr, "The OpenGL hardware renderer is slow on AMD GPUs due to an inefficient driver. Check out the link below for further information.\n"
```
Comment (FlatOutPS2) [inline plugins/GSdx/GLLoader.cpp]: It looks better if you replace the space between the sentences with `\n`.

**Comment by lightningterror** [comment]
All good now?

**Comment by FlatOutPS2** [comment]
As far as I'm concerned it can be merged now.

**Pull request merged**